File: swift-bridge/Sources/SpriteKitBridge/VideoNode.swift
import AVFoundation
import SpriteKit

@_cdecl("sk_video_node_new")
public func sk_video_node_new() -> UnsafeMutableRawPointer? {
    skRetain(SKVideoNode(avPlayer: AVPlayer()))
}

@_cdecl("sk_video_node_play")
public func sk_video_node_play(_ nodeHandle: UnsafeMutableRawPointer?) {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return }
    node.play()
}

@_cdecl("sk_video_node_pause")
public func sk_video_node_pause(_ nodeHandle: UnsafeMutableRawPointer?) {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return }
    node.pause()
}

@_cdecl("sk_video_node_get_size_w")
public func sk_video_node_get_size_w(_ nodeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return 0 }
    return Double(node.size.width)
}

@_cdecl("sk_video_node_get_size_h")
public func sk_video_node_get_size_h(_ nodeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return 0 }
    return Double(node.size.height)
}

@_cdecl("sk_video_node_set_size")
public func sk_video_node_set_size(_ nodeHandle: UnsafeMutableRawPointer?, _ width: Double, _ height: Double) {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return }
    node.size = CGSize(width: width, height: height)
}

@_cdecl("sk_video_node_get_anchor_x")
public func sk_video_node_get_anchor_x(_ nodeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return 0.5 }
    return Double(node.anchorPoint.x)
}

@_cdecl("sk_video_node_get_anchor_y")
public func sk_video_node_get_anchor_y(_ nodeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return 0.5 }
    return Double(node.anchorPoint.y)
}

@_cdecl("sk_video_node_set_anchor_point")
public func sk_video_node_set_anchor_point(_ nodeHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) {
    guard let node: SKVideoNode = skBorrow(nodeHandle) else { return }
    node.anchorPoint = CGPoint(x: x, y: y)
}

File: src/video_node.rs
use apple_cf::cg::{CGPoint, CGSize};

use crate::ffi;
use crate::node::AsNode;
use crate::private::handle_type;

handle_type!(VideoNode);

impl AsNode for VideoNode {
    fn as_node_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl VideoNode {
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_video_node_new()) }
    }

    pub fn play(&self) {
        unsafe { ffi::sk_video_node_play(self.ptr) };
    }

    pub fn pause(&self) {
        unsafe { ffi::sk_video_node_pause(self.ptr) };
    }

    #[must_use]
    pub fn size(&self) -> CGSize {
        CGSize::new(unsafe { ffi::sk_video_node_get_size_w(self.ptr) }, unsafe {
            ffi::sk_video_node_get_size_h(self.ptr)
        })
    }

    pub fn set_size(&self, size: CGSize) {
        unsafe { ffi::sk_video_node_set_size(self.ptr, size.width, size.height) };
    }

    #[must_use]
    pub fn anchor_point(&self) -> CGPoint {
        CGPoint::new(
            unsafe { ffi::sk_video_node_get_anchor_x(self.ptr) },
            unsafe { ffi::sk_video_node_get_anchor_y(self.ptr) },
        )
    }

    pub fn set_anchor_point(&self, anchor: CGPoint) {
        unsafe { ffi::sk_video_node_set_anchor_point(self.ptr, anchor.x, anchor.y) };
    }
}

impl Default for VideoNode {
    fn default() -> Self {
        Self::new().expect("SKVideoNode creation failed")
    }
}
